packages/teams-js/src/private/externalAppCardActionsForCEA.ts

```diff
@@ -0,0 +1,99 @@
+export namespace externalAppCardActionsForCEA {
+  /**
+   * @beta
+   * @hidden
+   * Delegates an Adaptive Card Action.OpenUrl request to the host for the application with the provided app ID.
+   * @internal
+   * Limited to Microsoft-internal use
+   * @param appId ID of the application the request is intended for. This must be a UUID
+   * @param conversationId To tell the bot what conversation the calls are coming from
+   * @param url The URL to open
+   * @returns Promise that resolves to ActionOpenUrlType indicating the type of URL that was opened on success and rejects with ActionOpenUrlError if the request fails
+   */
+  export function processActionOpenUrl(appId: string, conversationId: string, url: URL): Promise<ActionOpenUrlType> {
+    ensureInitialized(runtime, FrameContexts.content);
+
+    if (!isSupported()) {
+      throw errorNotSupportedOnPlatform;
+    }
+    validateId(appId, new Error('App id is not valid.'));
+    validateId(conversationId, new Error('conversation id is not valid.'));
+    return sendMessageToParentAsync<[ActionOpenUrlError, ActionOpenUrlType]>(
+      getApiVersionTag(
+        externalAppCardActionsTelemetryVersionNumber,
+        ApiName.ExternalAppCardActionsForCEA_ProcessActionOpenUrl,
+      ),
+      'externalAppCardActionsForCEA.processActionOpenUrl',
+      [appId, url.href, conversationId],
+    ).then(([error, response]: [ActionOpenUrlError, ActionOpenUrlType]) => {
+      if (error) {
+        throw error;
+      } else {
+        return response;
+      }
+    });
+  }
+
+  /**
+   * @beta
+   * @hidden
+   * Delegates an Adaptive Card Action.Submit request to the host for the application with the provided app ID
+   * @internal
+   * Limited to Microsoft-internal use
+   * @param appId ID of the application the request is intended for. This must be a UUID
+   * @param conversationId To tell the bot what conversation the calls are coming from
+   * @param actionSubmitPayload The Adaptive Card Action.Submit payload
+   * @returns Promise that resolves when the request is completed and rejects with ActionSubmitError if the request fails
+   */
+  export function processActionSubmit(
+    appId: string,
+    conversationId: string,
+    actionSubmitPayload: IAdaptiveCardActionSubmit,
+  ): Promise<void> {
+    ensureInitialized(runtime, FrameContexts.content);
+
+    if (!isSupported()) {
+      throw errorNotSupportedOnPlatform;
+    }
+    validateId(appId, new Error('App id is not valid.'));
+    validateId(conversationId, new Error('conversation id is not valid.'));
+    return sendMessageToParentAsync<[boolean, ActionSubmitError]>(
+      getApiVersionTag(
+        externalAppCardActionsTelemetryVersionNumber,
+        ApiName.ExternalAppCardActionsForCEA_ProcessActionSubmit,
+      ),
+      'externalAppCardActionsForCEA.processActionSubmit',
+      [appId, conversationId, actionSubmitPayload],
+    ).then(([wasSuccessful, error]: [boolean, ActionSubmitError]) => {
```

Comment:
>[wasSuccessful, error] [](http://example.com/codeflow?start=12&length=22)

Does this function need both a boolean and error return value pair? Could it just return only the error object and if the error object was undefined or null, that would mean "no error" ?

With this definition, you allow for the boolean to be `true` and the error object to be present, which would seem like an invalid state that you wouldn't want the code to allow.